### PR TITLE
HW-55341-Fixed bug to show all decimal values in fund transfer confirm page.

### PR DESCRIPTION
**Pull request by sarun-altmtrk**

### CODE REVIEW

**Comment by sarun-altmtrk** [comment]
Fixed transfer Confirmation Page is Missing the Currencies, also in this PR.
https://jira.site1.hyperwallet.local/browse/HW-55339

File: Common/Sources/Extensions/String.swift [hunk 1]
```diff
@@ -79,6 +79,30 @@ public extension String {
...
+    /// - Parameter currencyCode: the currency code
+    /// - Returns: a formatted String amount with currency code
+    func formatWithActualDecimal(with currencyCode: String?) -> String {
+        if !self.isEmpty {
+            let currencyFormatter = NumberFormatter()
+            let locale = Locale(identifier: Locale.preferredLanguages[0])
+            let localizedDecimalSeparator: Character = locale.decimalSeparator?.first ?? "."
+            let components = self.split(separator: localizedDecimalSeparator)
+            let fractionDigits = components.count <= 1 ? 2 : components[1].count
+            currencyFormatter.maximumFractionDigits = fractionDigits
```
Comment (ddhar-hw) [inline Common/Sources/Extensions/String.swift]: Please move the currency formatter logic to a seperate `func`

File: Common/Sources/Extensions/String.swift [hunk 2]
```diff
@@ -79,6 +79,30 @@ public extension String {
...
+            currencyFormatter.minimumFractionDigits = fractionDigits
+            currencyFormatter.minimumIntegerDigits = 1
+            currencyFormatter.numberStyle = .currency
+            currencyFormatter.currencyCode = currencyCode
+            currencyFormatter.currencySymbol = ""
+            let formattedAmountInNumber = currencyFormatter.number(from: self)
+            return currencyFormatter.string(for: formattedAmountInNumber) ?? ""
+        } else {
+            return ""
```
Comment (ddhar-hw) [inline Common/Sources/Extensions/String.swift]: You could return a blank string on the first line of the method with a ternary operator

File: Common/Sources/Extensions/String.swift [hunk 3]
```diff
@@ -79,6 +79,30 @@ public extension String {
...
+    /// Format an amount to a currency format with currency code
+    ///
+    /// - Parameter currencyCode: the currency code
+    /// - Returns: a formatted String amount with currency code
+    func formatWithActualDecimal(with currencyCode: String?) -> String {
+        if !self.isEmpty {
+            let currencyFormatter = NumberFormatter()
+            let locale = Locale(identifier: Locale.preferredLanguages[0])
+            let localizedDecimalSeparator: Character = locale.decimalSeparator?.first ?? "."
+            let components = self.split(separator: localizedDecimalSeparator)
```
Comment (ymoskovko-epam) [inline Common/Sources/Extensions/String.swift]: we don't use locale and localization in the Number formatters.

File: Common/Sources/Extensions/String.swift (hunk 2 above)
Comment (ymoskovko-epam) [reply Common/Sources/Extensions/String.swift]: or use `guard` statement

File: Transfer/Sources/ScheduleTransferSectionData.swift [hunk 4]
```diff
@@ -67,14 +67,17 @@ struct ScheduleTransferForeignExchangeData: ScheduleTransferSectionData {
                 let destinationCurrency = foreignExchange.destinationCurrency,
                 let rate = foreignExchange.rate {
                 let sourceAmountFormatted = String(format: "%@ %@",
-                                                   sourceAmount.format(with: sourceCurrency),
+                                                   sourceAmount.formatWithActualDecimal(with: sourceCurrency),
```
Comment (ymoskovko-epam) [inline Transfer/Sources/ScheduleTransferSectionData.swift]: I'm not sure about the naming of this method `formatWithActualDecimal(with:)`

Comment (ymoskovko-epam) [reply Transfer/Sources/ScheduleTransferSectionData.swift]: maybe `format(withActualDecimals:)` or `format(similarTo:)`

File: Common/Sources/Extensions/String.swift (hunk 1 above)
Comment (sarun-altmtrk) [reply Common/Sources/Extensions/String.swift]: Now following code has been removed.

File: Common/Sources/Extensions/String.swift (hunk 2 above)
Comment (sarun-altmtrk) [reply Common/Sources/Extensions/String.swift]: It has been written as existing function.

File: Common/Sources/Extensions/String.swift (hunk 3 above)
Comment (sarun-altmtrk) [reply Common/Sources/Extensions/String.swift]: Code has been modified.

File: Transfer/Sources/ScheduleTransferSectionData.swift (hunk 4 above)
Comment (sarun-altmtrk) [reply Transfer/Sources/ScheduleTransferSectionData.swift]: Function name has been changed.

File: Common/Sources/Extensions/String.swift [hunk 5]
```diff
@@ -79,6 +79,28 @@ public extension String {
         }
     }
 
+    /// Format an amount to a currency format with currency code
+    ///
+    /// - Parameter currencyCode: the currency code
+    /// - Returns: a formatted String amount with currency code
+    func format(withActualDecimal currencyCode: String?) -> String {
```
Comment (jsaini-hw) [inline Common/Sources/Extensions/String.swift]: No need to add this new method. Kindly Remove it

File: Transfer/Sources/ScheduleTransferSectionData.swift [hunk 6]
```diff
@@ -67,14 +67,17 @@ struct ScheduleTransferForeignExchangeData: ScheduleTransferSectionData {
                 let destinationCurrency = foreignExchange.destinationCurrency,
                 let rate = foreignExchange.rate {
                 let sourceAmountFormatted = String(format: "%@ %@",
-                                                   sourceAmount.format(with: sourceCurrency),
+                                                   sourceAmount.format(withActualDecimal: sourceCurrency),
```
Comment (jsaini-hw) [inline Transfer/Sources/ScheduleTransferSectionData.swift]: API returns formatted response so no need to format any value in this class. SO remove format from here. Just leave it as `String(format: "%@ %@",
                                                   sourceAmount,
                                                   sourceCurrency)`

File: Transfer/Sources/ScheduleTransferSectionData.swift [hunk 7]
```diff
@@ -67,14 +67,17 @@ struct ScheduleTransferForeignExchangeData: ScheduleTransferSectionData {
...
                                                    sourceCurrency)
                 let destinationAmountFormatted = String(format: "%@ %@",
-                                                        destinationAmount.format(with: destinationCurrency),
+                                                        destinationAmount
```
Comment (jsaini-hw) [inline Transfer/Sources/ScheduleTransferSectionData.swift]: API returns formatted response so no need to format any value in this class. SO remove format from here. Just leave it as `String(format: "%@ %@",
                                                        destinationAmount,
                                                        destinationCurrency)`

File: Transfer/Sources/ScheduleTransferSectionData.swift [hunk 8]
```diff
@@ -67,14 +67,17 @@ struct ScheduleTransferForeignExchangeData: ScheduleTransferSectionData {
...
+                                                            .format(withActualDecimal: destinationCurrency),
                                                         destinationCurrency)
-                let rateFormatted = String(format: "%@ = %@",
-                                           "1".format(with: sourceCurrency),
-                                           rate.format(with: destinationCurrency))
+                let rateFormatted = String(format: "%@ %@ = %@ %@",
```
Comment (jsaini-hw) [inline Transfer/Sources/ScheduleTransferSectionData.swift]: API returns formatted response so no need to format any value in this class. SO remove format from here. Just leave it as  `String(format: "%@ %@ = %@ %@",
                                           "1".format(with: sourceCurrency),
                                           sourceCurrency,
                                           rate,
                                           destinationCurrency)`

File: Transfer/Sources/ScheduleTransferSectionData.swift [hunk 9]
```diff
@@ -96,7 +99,7 @@ struct ScheduleTransferSummaryData: ScheduleTransferSectionData {
     init(transfer: HyperwalletTransfer) {
         if let destinationAmount = transfer.destinationAmount,
             let destinationCurrency = transfer.destinationCurrency {
-            let destinationAmountFormattedString = destinationAmount.format(with: destinationCurrency)
+            let destinationAmountFormattedString = destinationAmount.format(withActualDecimal: destinationCurrency)
```
Comment (jsaini-hw) [inline Transfer/Sources/ScheduleTransferSectionData.swift]: remove this. Not required

File: Transfer/Sources/ScheduleTransferSectionData.swift [hunk 10]
```diff
@@ -109,11 +112,18 @@ struct ScheduleTransferSummaryData: ScheduleTransferSectionData {
             let grossTransferAmount = transferAmountFormattedDouble + feeAmountFormattedDouble
 
             rows.append((title: "transfer_amount_confirmation".localized(),
-                         value: String(grossTransferAmount).format(with: destinationCurrency)))
-            rows.append((title: "transfer_fee_confirmation".localized(),
-                         value: feeAmount.format(with: destinationCurrency)))
+                         value: String(format: "%@ %@",
+                                       String(grossTransferAmount)
+                            .format(withActualDecimal: destinationCurrency), destinationCurrency)))
+            if feeAmount != "0.00" {
```
Comment (jsaini-hw) [inline Transfer/Sources/ScheduleTransferSectionData.swift]: move this condition to `guard let feeAmount = transfer.destinationFeeAmount, Double(feeAmount) != 0 else {...} `

File: Transfer/Sources/ScheduleTransferSectionData.swift [hunk 11]
```diff
@@ -109,11 +112,18 @@ struct ScheduleTransferSummaryData: ScheduleTransferSectionData {
...
+                rows.append((title: "transfer_fee_confirmation".localized(),
```
Comment (jsaini-hw) [inline Transfer/Sources/ScheduleTransferSectionData.swift]: API returns formatted response so no need to format any value in this class. SO remove format from here. Just leave it as `String(format: "%@ %@",
                                       feeAmount,
                                       destinationCurrency))`

File: Transfer/Sources/ScheduleTransferSectionData.swift [hunk 12]
```diff
@@ -109,11 +112,18 @@ struct ScheduleTransferSummaryData: ScheduleTransferSectionData {
...
+                         value: String(format: "%@ %@",
+                                       feeAmount
+                            .format(withActualDecimal: destinationCurrency),
+                                       destinationCurrency)))
+            }
             rows.append((title: "transfer_net_amount_confirmation".localized(),
-                         value: destinationAmountFormattedString))
+                         value: String(format: "%@ %@", destinationAmountFormattedString, destinationCurrency)))
```
Comment (jsaini-hw) [inline Transfer/Sources/ScheduleTransferSectionData.swift]: API returns formatted response so no need to format any value in this class. SO remove format from here. Just leave it as `String(format: "%@ %@", destinationAmount, destinationCurrency)`

Comment (jsaini-hw) [reply Transfer/Sources/ScheduleTransferSectionData.swift]: Also remove extra ()

File: Transfer/Sources/ScheduleTransferSectionData.swift (hunk 11 above)
Comment (jsaini-hw) [reply Transfer/Sources/ScheduleTransferSectionData.swift]: Also remove extra ()

File: Transfer/Sources/ScheduleTransferSectionData.swift [hunk 13]
```diff
@@ -109,11 +112,18 @@ struct ScheduleTransferSummaryData: ScheduleTransferSectionData {
             let grossTransferAmount = transferAmountFormattedDouble + feeAmountFormattedDouble
 
             rows.append((title: "transfer_amount_confirmation".localized(),
```
Comment (jsaini-hw) [inline Transfer/Sources/ScheduleTransferSectionData.swift]: Also remove extra ()

File: Transfer/Sources/ScheduleTransferSectionData.swift [hunk 14]
```diff
@@ -67,14 +67,17 @@ struct ScheduleTransferForeignExchangeData: ScheduleTransferSectionData {
...
-                let rateFormatted = String(format: "%@ = %@",
-                                           "1".format(with: sourceCurrency),
-                                           rate.format(with: destinationCurrency))
+                let rateFormatted = String(format: "%@ %@ = %@ %@",
+                                           "1".format(withActualDecimal: sourceCurrency),
+                                           sourceCurrency,
+                                           rate.format(withActualDecimal: destinationCurrency),
+                                           destinationCurrency)
 
                 rows.append((title: "transfer_fx_sell_confirmation".localized(), value:sourceAmountFormatted))
```
Comment (jsaini-hw) [inline Transfer/Sources/ScheduleTransferSectionData.swift]: Also remove extra ()

File: Transfer/Sources/ScheduleTransferSectionData.swift [hunk 15]
```diff
@@ -67,14 +67,17 @@ struct ScheduleTransferForeignExchangeData: ScheduleTransferSectionData {
...
                 rows.append((title: "transfer_fx_buy_confirmation".localized(), value: destinationAmountFormatted))
```
Comment (jsaini-hw) [inline Transfer/Sources/ScheduleTransferSectionData.swift]: Also remove extra ()

Comment (jsaini-hw) [reply Transfer/Sources/ScheduleTransferSectionData.swift]: Also remove `currencyFormatter.minimumFractionDigits = 2` from both func format(with currencyCode: String?) -> String and func formatToDouble(with currencyCode: String? = nil) -> Double functions

File: Transfer/Sources/ScheduleTransferSectionData.swift [hunk 16]
```diff
@@ -70,11 +70,14 @@ struct ScheduleTransferForeignExchangeData: ScheduleTransferSectionData {
                                                    sourceAmount.format(with: sourceCurrency),
                                                    sourceCurrency)
                 let destinationAmountFormatted = String(format: "%@ %@",
-                                                        destinationAmount.format(with: destinationCurrency),
+                                                        destinationAmount
+                                                        .format(with: destinationCurrency),
```
Comment (jsaini-hw) [inline Transfer/Sources/ScheduleTransferSectionData.swift]: remove this format as per my above comment

File: Transfer/Sources/ScheduleTransferSectionData.swift [hunk 17]
```diff
@@ -70,11 +70,14 @@ struct ScheduleTransferForeignExchangeData: ScheduleTransferSectionData {
...
                                                         destinationCurrency)
-                let rateFormatted = String(format: "%@ = %@",
+                let rateFormatted = String(format: "%@ %@ = %@ %@",
                                            "1".format(with: sourceCurrency),
-                                           rate.format(with: destinationCurrency))
+                                           sourceCurrency,
+                                           rate.format(with: destinationCurrency),
```
Comment (jsaini-hw) [inline Transfer/Sources/ScheduleTransferSectionData.swift]: why format again? In my previous review comment I asked to remove the format

File: Transfer/Sources/ScheduleTransferSectionData.swift [hunk 18]
```diff
@@ -109,11 +112,17 @@ struct ScheduleTransferSummaryData: ScheduleTransferSectionData {
...
 
             rows.append((title: "transfer_amount_confirmation".localized(),
-                         value: String(grossTransferAmount).format(with: destinationCurrency)))
+                         value: String(format: "%@ %@",
+                                       String(grossTransferAmount)
+                            .format(with: destinationCurrency), destinationCurrency)))
             rows.append((title: "transfer_fee_confirmation".localized(),
-                         value: feeAmount.format(with: destinationCurrency)))
+                         value: String(format: "%@ %@",
+                                       feeAmount.format(with: destinationCurrency),
```
Comment (jsaini-hw) [inline Transfer/Sources/ScheduleTransferSectionData.swift]: No need to format this as per my previous review comment

File: Transfer/Sources/ScheduleTransferSectionData.swift [hunk 19]
```diff
@@ -97,7 +100,7 @@ struct ScheduleTransferSummaryData: ScheduleTransferSectionData {
         if let destinationAmount = transfer.destinationAmount,
             let destinationCurrency = transfer.destinationCurrency {
             let destinationAmountFormattedString = destinationAmount.format(with: destinationCurrency)
```
Comment (jsaini-hw) [inline Transfer/Sources/ScheduleTransferSectionData.swift]: No need to format this as per my previous review comment. Remove this

File: Transfer/Sources/ScheduleTransferSectionData.swift [hunk 20]
```diff
@@ -97,7 +100,7 @@ struct ScheduleTransferSummaryData: ScheduleTransferSectionData {
...
-            guard let feeAmount = transfer.destinationFeeAmount else {
+            guard let feeAmount = transfer.destinationFeeAmount, Double(feeAmount) != 0 else {
                 rows.append((title: "transfer_amount_confirmation".localized(),
                              value: destinationAmountFormattedString))
```
Comment (jsaini-hw) [inline Transfer/Sources/ScheduleTransferSectionData.swift]: No need to format this as per my previous review comment, so use destinationAmount instead

File: Transfer/Sources/ScheduleTransferSectionData.swift [hunk 21]
```diff
@@ -109,11 +112,17 @@ struct ScheduleTransferSummaryData: ScheduleTransferSectionData {
...
+                            .format(with: destinationCurrency), destinationCurrency)))
             rows.append((title: "transfer_fee_confirmation".localized(),
-                         value: feeAmount.format(with: destinationCurrency)))
+                         value: String(format: "%@ %@",
+                                       feeAmount.format(with: destinationCurrency),
+                                       destinationCurrency)))
             rows.append((title: "transfer_net_amount_confirmation".localized(),
-                         value: destinationAmountFormattedString))
+                         value: String(format: "%@ %@",
+                                       destinationAmountFormattedString,
```
Comment (jsaini-hw) [inline Transfer/Sources/ScheduleTransferSectionData.swift]: No need to format this as per my previous review comment, so use destinationAmount instead

File: Transfer/Sources/ScheduleTransferSectionData.swift (hunk 20 above)
Comment (jsaini-hw) [reply Transfer/Sources/ScheduleTransferSectionData.swift]: Remove the below line as well. As per ticket when there's no fee do not show fee and You will Receive(transfer_net_amount_confirmation) cell. Only transfer_amount_confirmation will be shown

File: Transfer/Sources/ScheduleTransferSectionData.swift [hunk 22]
```diff
@@ -66,12 +66,9 @@ struct ScheduleTransferForeignExchangeData: ScheduleTransferSectionData {
                 let destinationAmount = foreignExchange.destinationAmount,
                 let destinationCurrency = foreignExchange.destinationCurrency,
                 let rate = foreignExchange.rate {
-                let sourceAmountFormatted = String(format: "%@ %@",
-                                                   sourceAmount.format(with: sourceCurrency),
-                                                   sourceCurrency)
-                let destinationAmountFormatted = String(format: "%@ %@",
-                                                        destinationAmount.format(with: destinationCurrency),
-                                                        destinationCurrency)
+                let sourceAmountFormatted = sourceAmount.format(with: sourceCurrency)
```
Comment (jsaini-hw) [inline Transfer/Sources/ScheduleTransferSectionData.swift]: @sarun-altmtrk  no need to format the sourceAmount. So update it to `let sourceAmountFormatted = String(format: "%@ %@", sourceAmount, sourceCurrency)`

File: Transfer/Sources/ScheduleTransferSectionData.swift [hunk 23]
```diff
@@ -66,12 +66,9 @@ struct ScheduleTransferForeignExchangeData: ScheduleTransferSectionData {
...
+                let destinationAmountFormatted = destinationAmount.format(with: destinationCurrency)
```
Comment (jsaini-hw) [inline Transfer/Sources/ScheduleTransferSectionData.swift]: Arun no need to format the sourceAmount. So update it to `let destinationAmountFormatted = String(format: "%@ %@", destinationAmount, destinationCurrency)`

File: Transfer/Sources/ScheduleTransferSectionData.swift [hunk 24]
```diff
@@ -97,11 +94,9 @@ struct ScheduleTransferSummaryData: ScheduleTransferSectionData {
         if let destinationAmount = transfer.destinationAmount,
             let destinationCurrency = transfer.destinationCurrency {
             let destinationAmountFormattedString = destinationAmount.format(with: destinationCurrency)
-            guard let feeAmount = transfer.destinationFeeAmount else {
+            guard let feeAmount = transfer.destinationFeeAmount, Double(feeAmount) != 0 else {
                 rows.append((title: "transfer_amount_confirmation".localized(),
```
Comment (jsaini-hw) [inline Transfer/Sources/ScheduleTransferSectionData.swift]: remove extra (

File: Transfer/Sources/ScheduleTransferSectionData.swift (hunk 20 above)
Comment (jsaini-hw) [reply Transfer/Sources/ScheduleTransferSectionData.swift]: @sarun-altmtrk Kindly fix the above comment. `No need to format this as per my previous review comment, so use destinationAmount instead`

File: Transfer/Sources/ScheduleTransferSectionData.swift (hunk 19 above)
Comment (jsaini-hw) [reply Transfer/Sources/ScheduleTransferSectionData.swift]: @sarun-altmtrk Kindly fix the above review comment

File: Transfer/Sources/ScheduleTransferSectionData.swift [hunk 25]
```diff
@@ -113,7 +108,7 @@ struct ScheduleTransferSummaryData: ScheduleTransferSectionData {
             rows.append((title: "transfer_fee_confirmation".localized(),
                          value: feeAmount.format(with: destinationCurrency)))
             rows.append((title: "transfer_net_amount_confirmation".localized(),
-                         value: destinationAmountFormattedString))
+                         value:destinationAmountFormattedString))
```
Comment (jsaini-hw) [inline Transfer/Sources/ScheduleTransferSectionData.swift]: As I mentioned in the previous comment delete destinationAmountFormattedString and use destinationAmount instead. Also remove extra (). Also do not remove space after value:

File: Transfer/Sources/ScheduleTransferSectionData.swift [hunk 26]
```diff
@@ -66,17 +66,14 @@ struct ScheduleTransferForeignExchangeData: ScheduleTransferSectionData {
...
-                                                        destinationAmount.format(with: destinationCurrency),
-                                                        destinationCurrency)
-                let rateFormatted = String(format: "%@ = %@",
+                let sourceAmountFormatted = String(format: "%@ %@", sourceAmount, sourceCurrency)
+                let destinationAmountFormatted = String(format: "%@ %@", destinationAmount, destinationCurrency)
+
+                let rateFormatted = String(format: "%@ %@ = %@ %@",
                                            "1".format(with: sourceCurrency),
-                                           rate.format(with: destinationCurrency))
+                                           sourceCurrency, rate, destinationCurrency)
```
Comment (jsaini-hw) [inline Transfer/Sources/ScheduleTransferSectionData.swift]: Fix the swiftlint warning

File: Transfer/Sources/ScheduleTransferSectionData.swift [hunk 27]
```diff
@@ -66,17 +66,14 @@ struct ScheduleTransferForeignExchangeData: ScheduleTransferSectionData {
...
-                                                   sourceCurrency)
-                let destinationAmountFormatted = String(format: "%@ %@",
-                                                        destinationAmount.format(with: destinationCurrency),
-                                                        destinationCurrency)
-                let rateFormatted = String(format: "%@ = %@",
+                let sourceAmountFormatted = String(format: "%@ %@", sourceAmount, sourceCurrency)
+                let destinationAmountFormatted = String(format: "%@ %@", destinationAmount, destinationCurrency)
+
+                let rateFormatted = String(format: "%@ %@ = %@ %@",
                                            "1".format(with: sourceCurrency),
```
Comment (jsaini-hw) [inline Transfer/Sources/ScheduleTransferSectionData.swift]: remove format from `"1".format(with: sourceCurrency) ` as well.

File: Transfer/Sources/ScheduleTransferSectionData.swift [hunk 28]
```diff
@@ -113,7 +107,7 @@ struct ScheduleTransferSummaryData: ScheduleTransferSectionData {
             rows.append((title: "transfer_fee_confirmation".localized(),
                          value: feeAmount.format(with: destinationCurrency)))
```
Comment (jsaini-hw) [inline Transfer/Sources/ScheduleTransferSectionData.swift]: why fee amount is still being formatted? remove `format(with: destinationCurrency)` from feeAmount

File: Transfer/Sources/ScheduleTransferSectionData.swift [hunk 29]
```diff
@@ -96,24 +91,21 @@ struct ScheduleTransferSummaryData: ScheduleTransferSectionData {
...
+                             value: destinationAmount))
                 return
             }
             let transferAmountFormattedDouble = destinationAmount.formatToDouble(with: destinationCurrency)
             let feeAmountFormattedDouble = feeAmount.formatToDouble(with: destinationCurrency)
             let grossTransferAmount = transferAmountFormattedDouble + feeAmountFormattedDouble
 
             rows.append((title: "transfer_amount_confirmation".localized(),
-                         value: String(grossTransferAmount).format(with: destinationCurrency)))
+                         value: String(grossTransferAmount)))
```
Comment (jsaini-hw) [inline Transfer/Sources/ScheduleTransferSectionData.swift]: grossAmount still needs to be formatted as it's calculated and not returned from API. So add ` String(grossTransferAmount).format(with: destinationCurrency) `

**Pull request merged**